File: cynq/mmio.cpp
#include "mmio.h"

#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>

// map a udma buffer
mapped_device mmioGetMmap(std::string dev, long addr, long size) {
  mapped_device mmio;
  mmio.size = size;
  mmio.addr = addr;
  if ((mmio.fd  = open(dev.c_str(), O_SYNC|O_RDWR)) != -1)
    mmio.mmap = (char*) mmap(NULL, mmio.size, PROT_READ|PROT_WRITE, MAP_SHARED, mmio.fd, mmio.addr);
  return mmio;
}

// free a mapped buffer
void mmioFreeMmap(mapped_device to_free) {
  munmap(to_free.mmap, to_free.size);
  close(to_free.fd);
}

File: ponq/mmio.py
import mmap, os, numpy as np, sys

class mmio:
  def __init__(self, reqbase):
    self.reqbase = int(reqbase)
    # self.reqrange = int(reqrange) % mmap.PAGESIZE
    self.reqrange = mmap.PAGESIZE

    # align requested range to page boundaries
    #self.base = reqbase & ~(mmap.PAGESIZE - 1)  # chops off sub-page bits
    #self.shift = self.reqbase - self.base       # calculates window shift
    #self.range = self.shift + self.reqrange     # adds shift value to range

    self.devmem = os.open("/dev/mem", os.O_RDWR | os.O_SYNC)
    self.memmap = mmap.mmap(self.devmem, self.reqrange, mmap.MAP_SHARED,
        mmap.PROT_READ | mmap.PROT_WRITE, offset=self.reqbase)

    # create numpy uint32 array from buffer
    # self.array = np.frombuffer(self.memmap, np.uint32, int(self.reqrange/4))

  def write(self, address, value):
    byteval = value.to_bytes(4, byteorder=sys.byteorder)
    self.memmap[address*4:(address+1)*4] = byteval

  def read(self, address):
    byteval = self.memmap[address*4:(address+1)*4]
    value = int.from_bytes(byteval, byteorder=sys.byteorder)
    return value

  def close(self):
    self.memmap.close()
    os.close(self.devmem)
